astropy/coordinates/spectral_coordinate.py

```diff
@@ -0,0 +1,842 @@
+            The target frame or coordinate.
+
+        Returns
+        -------
+        pos_hat : `BaseRepresentation`
+            Position representation.
+        """
+        d_pos = (target.data.without_differentials() -
+                 observer.data.without_differentials()).to_cartesian()
+
+        dp_norm = d_pos.norm()
+
+        # Reset any that are 0 to 1 to avoid nans from 0/0
+        dp_norm.ravel()[dp_norm.ravel() == 0] = 1 * dp_norm.unit
+
+        pos_hat = d_pos / dp_norm
+
+        return pos_hat
+
+    def _change_observer_to(self, observer, target=None):
+        """
+        Moves the observer to the provided coordinate/frame.
+
+        Parameters
+        ----------
+        observer : `~astropy.coordinates.BaseCoordinateFrame` or `~astropy.coordinates.SkyCoord`
+            The new observation frame or coordinate.
+        target : `~astropy.coordinates.SkyCoord`, optional
+            The `~astropy.coordinates.SkyCoord` object representing the target of the observation.
+            If none given, defaults to currently defined target.
+
+        Returns
+        -------
+        new_coord : `SpectralCoord`
+            The new coordinate object representing the spectral data
+            transformed to the new observer frame.
+        """
+
+        if self.observer is None:
+            raise ValueError("No observer has been set, cannot change "
+                             "observer.")
+
+        target = self.target if target is None else target
+
+        # Check velocities and distance values on the new frame. This is
+        #  handled in the frame validation.
+        observer = self._validate_coordinate(observer)
+
+        # Calculate the initial and final los velocity
+        init_obs_vel = self._calculate_radial_velocity(self.observer, target)
+        fin_obs_vel = self._calculate_radial_velocity(observer, target)
+
+        line_of_sight_unit_vec = self._norm_d_pos(observer, target).xyz
+
+        new_data = self._project_velocity_and_shift(init_obs_vel, fin_obs_vel, line_of_sight_unit_vec)
+
+        new_coord = self._copy(value=new_data,
+                               observer=observer,
+                               target=target)
+
+        return new_coord
+
+    def _project_velocity_and_shift(self, init_vel, fin_vel, line_of_sight_unit_vec):
+        """
+        Calculate the velocity projection given two vectors.
+
+        Parameters
+        ----------
+        init_vel : `u.Quantity`
+            Initial velocity vector.
+        fin_vel : `u.Quantity`
+            Final velocity vector.
+
+        Returns
+        -------
+        new_data : `u.Quantity`
+            Spectral axis data with velocity shift applied.
+        """
+
+        # Project the velocity shift vector onto the line-of-sight vector
+        # between the target and the new observation frame.
+        init_proj_vel = np.dot(init_vel, line_of_sight_unit_vec) * line_of_sight_unit_vec
```

Comment:
If one keeps this in terms of representations, this is just `line_of_sight_unit_vec.dot(init_vel) * line_of sight_unit_vec` (and no need for `NUMPY_LT_1_17`)

Reply:
Done